sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -509,13 +509,28 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
     }
   }
 
+  private def convertTypedLiteralToString(literal: Literal): String = literal match {
+    case Literal(data: Int, _: DateType) =>
+      DateFormatter(getZoneId(SQLConf.get.sessionLocalTimeZone)).format(data)
+    case Literal(data: Long, _: TimestampType) =>
+      TimestampFormatter.getFractionFormatter(getZoneId(SQLConf.get.sessionLocalTimeZone))
+        .format(data)
+    case Literal(data: CalendarInterval, _: CalendarIntervalType) => data.toString
+    case Literal(data: Array[Byte], _: BinaryType) =>
+      UTF8String.fromBytes(data).toString
+    case Literal(_, dataType) =>
+      throw new IllegalArgumentException(s"Literals of type '$dataType' are not" +
+        " supported when use typed literal as partition col value.")
+  }
+
   /**
    * Convert a constant of any type into a string. This is typically used in DDL commands, and its
    * main purpose is to prevent slight differences due to back to back conversions i.e.:
    * String -> Literal -> String.
    */
   protected def visitStringConstant(ctx: ConstantContext): String = withOrigin(ctx) {
```

Comment:
can we support all the literals here? e.g.
```
expression(ctx) match {
  case l: Litral => if (l.value == null) null else Cast(l, StringType, timezone).eval()
  case _ => fail... // this should never happen
}
```

Reply:
> can we support all the literals here? e.g.
> 
> ```
> expression(ctx) match {
>   case l: Litral => if (l.value == null) null else Cast(l, StringType, timezone).eval()
>   case _ => fail... // this should never happen
> }
> ```

your mean support all `constant` here
```
constant
    : NULL                                                                                     #nullLiteral
    | interval                                                                                 #intervalLiteral
    | identifier STRING                                                                        #typeConstructor
    | number                                                                                   #numericLiteral
    | booleanValue                                                                             #booleanLiteral
    | STRING+                                                                                  #stringLiteral
    ;
```

and use `expression` to visit this constant then we don't need so many convert logic.

Reply:
> ... but Cast.eval can't run here.but Cast.eval can't run here.

Why? What's the error?

Reply:
> > ... but Cast.eval can't run here.
> 
> Why? What's the error?

SInce `eval()` need input, you mean write it like this ?
```
  protected def visitConstant(ctx: ConstantContext): String = withOrigin(ctx) {
    expression(ctx) match {
      case l: Literal => if (l.value == null) {
        null
      } else {
        Cast(l, StringType, Some(SQLConf.get.sessionLocalTimeZone))
          .eval(InternalRow.apply(l.value)).toString
      }
      case _ =>
        throw IllegalArgumentException("")
    }
  }
```

Reply:
Hwo about current change?